include/seastar/core/shared_future.hh

```diff
@@ -296,6 +310,11 @@ public:
     bool valid() const noexcept {
         return bool(_state);
     }
+
+    /// \cond internal
+    template<typename... U>
+    friend bool internal::has_scheduled_task(const shared_future<U...>& p) noexcept;
+    /// \endcond
```

Comment:
i'd suggest put all these tester probes into `futures_test.cc`, and only leave a friend declaration. and instead of put the helper into `internal` namespace, put it into `seastar`, so that we can just put
```c++
friend class shared_future_tester;
```
in `shared_future`. and have
```c++
namespace seastar {
class shared_future_tester {
public:
    template <typename... T>
    static bool has_scheduled_task(const shared_future<T...>& f) noexcept {
        return f._state->has_scheduled_task();
    }
};
}
``` 
in `futures_test.cc`. 

for couple reasons:

1. simpler this way. as it avoids the forward declarations
2. less distraction. so we put less test related stuff in the design under test
3. more obvious that the probe is a test-only facility
4. and ensures that the test-only facility can only be used by test.

Reply:
The `shared_future_tester` is a definitely a good idea, I'll do it.